eng/common/TestResources/New-TestResources.ps1

```diff
@@ -139,6 +139,122 @@ function BuildBicepFile([System.IO.FileSystemInfo] $file) {
     return $templateFilePath
 }
 
+function BuildDeploymentOutputs([string]$serviceDirectoryPrefix, [object]$azContext, [object]$deployment) {
+    # Add default values
+    $deploymentOutputs = [Ordered]@{
+        "${serviceDirectoryPrefix}CLIENT_ID" = $TestApplicationId;
+        "${serviceDirectoryPrefix}CLIENT_SECRET" = $TestApplicationSecret;
+        "${serviceDirectoryPrefix}TENANT_ID" = $azContext.Tenant.Id;
+        "${serviceDirectoryPrefix}SUBSCRIPTION_ID" =  $azContext.Subscription.Id;
+        "${serviceDirectoryPrefix}RESOURCE_GROUP" = $resourceGroup.ResourceGroupName;
+        "${serviceDirectoryPrefix}LOCATION" = $resourceGroup.Location;
+        "${serviceDirectoryPrefix}ENVIRONMENT" = $azContext.Environment.Name;
+        "${serviceDirectoryPrefix}AZURE_AUTHORITY_HOST" = $azContext.Environment.ActiveDirectoryAuthority;
+        "${serviceDirectoryPrefix}RESOURCE_MANAGER_URL" = $azContext.Environment.ResourceManagerUrl;
+        "${serviceDirectoryPrefix}SERVICE_MANAGEMENT_URL" = $azContext.Environment.ServiceManagementUrl;
+    }
+
+    MergeHashes $EnvironmentVariables $(Get-Variable deploymentOutputs)
+
+    foreach ($key in $deployment.Outputs.Keys) {
+        $variable = $deployment.Outputs[$key]
+
+        # Work around bug that makes the first few characters of environment variables be lowercase.
+        $key = $key.ToUpperInvariant()
+
+        if ($variable.Type -eq 'String' -or $variable.Type -eq 'SecureString') {
+            $deploymentOutputs[$key] = $variable.Value
+        }
+    }
+
+    return $deploymentOutputs
+}
+
+function ShouldMarkValueAsSecret([string]$serviceDirectoryPrefix, [string]$key, [string]$value, [array]$allowedValues) {
+    $logOutputNonSecret = @(
+        # Environment Variables
+        "RESOURCEGROUP_NAME",
+        # Deployment Outputs
+        "CLIENT_ID",
+        "TENANT_ID",
+        "SUBSCRIPTION_ID",
+        "RESOURCE_GROUP",
+        "LOCATION",
+        "ENVIRONMENT",
+        "AZURE_AUTHORITY_HOST",
+        "RESOURCE_MANAGER_URL",
+        "SERVICE_MANAGEMENT_URL"
+    )
+
+    $suffix1 = $key -replace $serviceDirectoryPrefix, ""
```

Comment:
Should we just do an ends with key?

Reply:
That would be a lot simpler!

My one concern is it's not quite exhaustive and values could slip through the cracks. For example if a template output contains `CONNECTION_STRING_FOR_ENVIRONMENT` or something, then this would match and we would print values that we want secret.

Reply:
We should default to assuming any output is a secret then. This is how the Azure core pipelines work: headers and parameters are secrets unless marked otherwise.

Reply:
Yes that's how I wrote it. To make a value public is opt-in otherwise if it doesn't match the "allowlist" then it's marked secret.